lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -892,14 +892,39 @@ def alerts(self) -> dict:
                 continue
             # Construct an ID based on what's in the alert rules
             error_messages = []
+            tool = CosTool(self._charm)
             for group in alert_rules["groups"]:
+
+                # Copy off rules so we don't modify an object we're iterating over
```

Comment:
Immediately below this. The next line.

Reply:
Yes, it is by reference. No, it should not be `.copy()`. The intention is not a "real" copy, it is because some Python versions (and some datastructures) will throw `RuntimeError: dictionary changed size during iteration` if an additional reference variable is not used and we ever, at any point, modify the length (`del`, `pop`, etc).

It's not extra bookkeeping for no reason.